src/baseTrie.ts

```diff
@@ -468,114 +392,100 @@ export class Trie {
+            for (const child of children) {
+              const keyExtension = child[0] as number[]
+              const childRef = child[1] as Buffer
               const childKey = key.concat(keyExtension)
               const priority = childKey.length
-              taskExecutor.execute(priority, (taskCallback: Function) => {
-                self._lookupNode(childRef, (e: Error, childNode: TrieNode) => {
-                  if (e) {
-                    return cb(e, node)
-                  }
-                  taskCallback()
-                  processNode(childRef, childNode, childKey, cb)
-                })
+              taskExecutor.execute(priority, async (taskCallback: Function) => {
+                const childNode = await self._lookupNode(childRef)
+                taskCallback()
+                if (childNode) {
+                  processNode(childRef, childNode as TrieNode, childKey)
+                }
               })
-            },
-            cb,
-          )
-        },
-        only: function (childIndex: number) {
-          if (!(node instanceof BranchNode)) {
-            return cb(new Error('Expected branch node'))
-          }
-          const childRef = node.getBranch(childIndex)
-          const childKey = key.slice()
-          childKey.push(childIndex)
-          const priority = childKey.length
-          taskExecutor.execute(priority, (taskCallback: Function) => {
-            self._lookupNode(childRef as Buffer, (e: Error, childNode: TrieNode) => {
-              if (e) {
-                return cb(e, node)
-              }
+            }
+          },
+          only: async (childIndex: number) => {
+            if (!(node instanceof BranchNode)) {
+              throw new Error('Expected branch node')
+            }
+            const childRef = node.getBranch(childIndex)
+            if (!childRef) {
+              throw new Error('Could not get branch of childIndex')
+            }
+            const childKey = key.slice()
+            childKey.push(childIndex)
+            const priority = childKey.length
+            taskExecutor.execute(priority, async (taskCallback: Function) => {
+              const childNode = await self._lookupNode(childRef)
               taskCallback()
-              processNode(childRef as Buffer, childNode, childKey, cb)
+              if (childNode) {
+                await processNode(childRef as Buffer, childNode, childKey)
+              } else {
+                // could not find child node
+                resolve()
+              }
             })
-          })
-        },
+          },
+        }
+
+        if (node) {
+          onNode(nodeRef, node, key, walkController)
+        } else {
+          resolve()
+        }
       }
 
-      onNode(nodeRef, node, key, walkController)
-    }
+      const node = await this._lookupNode(root)
+      if (node) {
+        await processNode(root, node as TrieNode, [])
+      } else {
+        resolve()
+      }
+    })
```

Comment:
Whew, this is a biggy. I have to say that I just can't follow the changes here in `_walkTrie`, eventually someone else can give this another look (//cc @s1na @evertonfraga) For the moment I am not getting the whole picture.

Reply:
Ok, I think I finally got the replacement constructs, can give this a go now too. 😀